doc/man/gnunet-core.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd April 11, 2014
.Dt GNUNET-CORE 1
.Os
.Sh NAME
.Nm gnunet-core
.Nd monitor CORE subsystem
.Sh SYNOPSIS
.Nm
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl I Ar HELLOURI | Fl -import-hello= Ns Ar HELLOURI
.Op Fl H | -export-hello
.Op Fl D | -dump-hellos
.Op Fl m | -monitor
.Op Fl v | -version
.Op Fl V | -verbose
.Sh DESCRIPTION
.Nm
is a tool to access various functions of GNUnet's core subsystem
from the command-line.
.Bl -tag -width indent
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Configuration file to use.
.It Fl h | -help
Print the help page.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Change the loglevel.
Possible values for LOGLEVEL are ERROR, WARNING, INFO and DEBUG.
.It Fl I Ar HELLOURI | Fl -import-hello= Ns Ar HELLOURI
Import a HELLO URI (into the peerstore and make it available to core).
.It Fl H | -export-hello
Print a freshly signed HELLO URI for our peer.
.It Fl D | -dump-hellos
Print all known HELLOs (peer IDs and addresses).
These are only all known, and not necessarily connected peers.
.It Fl m | -monitor
In monitor mode, gnunet-core will continuously print the connection status,
instead of giving just a snapshot.
.It Fl i | -show-identity
Output current local peer identity.
.It Fl s | -connection-status
Show current peer connection status.
.It Fl v | -version
Print the version number.
.It Fl V | -verbose
Be verbose.
.El
.\".Sh EXAMPLES
.Sh SEE ALSO
.Xr gnunet-transport 1
.Pp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


NEWS
v0.21.2:
  - Reworked TESTING API
  - Disabled many old-style tests
  - TESTING: introduce GNUNET_TESTING_command_new_ac
  - UTIL: change return type to 'bool' from GNUNET_is_zero()
  - Added GNUNET_GNS_parse_ztld helper API

v0.21.1:
  - Communicator and connectivity bugfixes

v0.21.0:
  - Reworked PEERSTORE API
  - Added record flag for maintenance records
  - ensure traits can be generated with subsystem-specific prefixes for the symbols
  - libgnunettesting first major testing NG refactor towards getting dependency structure streamlined
  - Remove single-use API macro GNUNET_VA_ARG_ENUM
  - major revision of blind signature API
  - Introduced closure to hold store context when caling function to add hello in peerstore.
  - Added DDLs for handling GNUNET_PEERSTORE_StoreHelloContext
  - Removed old hello functionality.
  - Refactoring components under src/ into lib/, plugin/, cli/ and service/
  - add support for encoding/decoding double values as part of JSON to libgnunetjson
  - Changed method GNUNET_HELLO_builder_get_expiration_time to not need parameter GNUNET_HELLO_Builder.
  - Code moved to the core package to get rid of circular dependencies.
  - Moved code to testing to have more generic test setup, which can be used not only from within transport.
  - The old hello design replaced by the new  hello design.
  - Added api to get notified when hellos are stored with peerstore service.
  - Added api to store hellos with peerstore service.
  - Changed new hello uri api to allow to change the expiration time
  - Moved start peer command to testing subsystem.
  - Removed all usage of old transport api, beside peerinfo tool, gnunet-transport cli and usage in transport layer itself.
  - Added __attribute__((deprecated)) to the old transport API

v0.20.0:
  - GNUNET_TESTING_get_testname_from_underscore renamed to GNUNET_STRINGS_get_suffix_from_binary_name and moved from libgnunettesting to libgnuneutil
  - Move GNUNET_s into libgnunetutil.
  - re-introduce compiler annotation for array size in signature
  - function-signature adjustment due to compiler error
  - GNUNET_PQ_get_oid removed, GNUNET_PQ_get_oid_by_name improved
  - Added GNUNET_PQ_get_oid_by_name
  - added GNUNET_PQ_get_oid()
  - Added new CCA-secure KEM and use in IDENTITY encryption
  - Add KEM API to avoid ephemeral private key management
  - Add new GNUNET_PQ_event_do_poll() API to gnunet_pq_lib.h
  - Added API to support arrays in query results
  - Improve PQ API documentation.
  - API for array types extended for times
  - API extended for array query types
  - relevant array-types in queries (not results) in postgresql added
  - just style fixes, int to enum
  - initial steps towards support of array-types in posgresql
  - adds GNUNET_JSON_spec_object_const() and GNUNET_JSON_spec_array_const()

v0.19.4:
  - No changes

v0.19.3:
  - We now detect MySQL's strange, version-dependent my_bool type on configure.
  - Add pkg-config definitions for gnunet messenger.

v0.19.2:
  - New changelog format origin


src/cli/core/gnunet-core.c
/*
     This file is part of GNUnet.
     Copyright (C) 2011, 2012, 2014 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file core/gnunet-core.c
 * @brief Print information about other peers known to CORE.
 * @author Nathan Evans
 */
#include "platform.h"
#include "gnunet_time_lib.h"
#include "gnunet_util_lib.h"
#include "gnunet_core_service.h"
#include "gnunet_util_lib.h"

/**
 * Return code
 */
static int ret;

/**
 * Option -m.
 */
static int monitor_connections;

/**
 * Option -i.
 */
static int show_pid;

/**
 * Option -s.
 */
static int show_conns;

/**
 * Handle to the CORE monitor.
 */
static struct GNUNET_CORE_MonitorHandle *mh;

/**
 * Peer private key
 */
static struct GNUNET_CRYPTO_EddsaPrivateKey my_private_key;

/**
 * Peer identity
 */
static struct GNUNET_PeerIdentity my_full_id;

/**
 * Task run in monitor mode when the user presses CTRL-C to abort.
 * Stops monitoring activity.
 *
 * @param cls NULL
 */
static void
shutdown_task (void *cls)
{
  (void) cls;
  if (NULL != mh)
  {
    GNUNET_CORE_monitor_stop (mh);
    mh = NULL;
  }
}


/**
 * Function called to notify core users that another
 * peer changed its state with us.
 *
 * @param cls closure
 * @param peer the peer that changed state
 * @param state new state of the peer
 * @param timeout timeout for the new state
 */
static void
monitor_cb (void *cls,
            const struct GNUNET_PeerIdentity *peer,
            enum GNUNET_CORE_KxState state,
            struct GNUNET_TIME_Absolute timeout)
{
  struct GNUNET_TIME_Absolute now = GNUNET_TIME_absolute_get ();
  const char *now_str;
  const char *state_str;

  (void) cls;
  if (((NULL == peer) || (GNUNET_CORE_KX_ITERATION_FINISHED == state)) &&
      (GNUNET_NO == monitor_connections))
  {
    GNUNET_SCHEDULER_shutdown ();
    return;
  }

  switch (state)
  {
  case GNUNET_CORE_KX_STATE_DOWN:
    /* should never happen, as we immediately send the key */
    state_str = _ ("fresh connection");
    break;

  case GNUNET_CORE_KX_STATE_KEY_SENT:
    state_str = _ ("key sent");
    break;

  case GNUNET_CORE_KX_STATE_KEY_RECEIVED:
    state_str = _ ("key received");
    break;

  case GNUNET_CORE_KX_STATE_UP:
    state_str = _ ("connection established");
    break;

  case GNUNET_CORE_KX_STATE_REKEY_SENT:
    state_str = _ ("rekeying");
    break;

  case GNUNET_CORE_KX_PEER_DISCONNECT:
    state_str = _ ("disconnected");
    break;

  case GNUNET_CORE_KX_ITERATION_FINISHED:
    return;

  case GNUNET_CORE_KX_CORE_DISCONNECT:
    fprintf (stderr,
             "%s\n",
             _ ("Connection to CORE service lost (reconnecting)"));
    return;

  default:
    state_str = _ ("unknown state");
    break;
  }
  now_str = GNUNET_STRINGS_absolute_time_to_string (now);
  fprintf (stdout,
           _ ("%24s: %-30s %4s (timeout in %6s)\n"),
           now_str,
           state_str,
           GNUNET_i2s (peer),
           GNUNET_STRINGS_relative_time_to_string (
             GNUNET_TIME_absolute_get_remaining (timeout),
             GNUNET_YES));
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  char *keyfile;
  (void) cls;
  (void) cfgfile;

  if (NULL != args[0])
  {
    fprintf (stderr, _ ("Invalid command line argument `%s'\n"), args[0]);
    return;
  }
  if (! show_pid && ! show_conns && ! monitor_connections)
  {
    fprintf (stderr, "%s", _ ("No argument given.\n"));
    ret = 1;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }
  GNUNET_SCHEDULER_add_shutdown (&shutdown_task, NULL);
  if (GNUNET_OK !=
      GNUNET_CONFIGURATION_get_value_filename (cfg,
                                               "PEER",
                                               "PRIVATE_KEY",
                                               &keyfile))
  {
    GNUNET_log (
      GNUNET_ERROR_TYPE_ERROR,
      _ ("Core service is lacking HOSTKEY configuration setting.  Exiting.\n"));
    GNUNET_SCHEDULER_shutdown ();
    ret =  1;
    return;
  }
  if (GNUNET_SYSERR ==
      GNUNET_CRYPTO_eddsa_key_from_file (keyfile,
                                         GNUNET_YES,
                                         &my_private_key))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "Failed to read peer's private key!\n");
    GNUNET_SCHEDULER_shutdown ();
    ret = 1;
    GNUNET_free (keyfile);
    return;
  }
  GNUNET_free (keyfile);
  GNUNET_CRYPTO_eddsa_key_get_public (&my_private_key, &my_full_id.public_key);
  if (show_pid)
    fprintf (stdout,
             _ ("Current local peer identity: %s\n"),
             GNUNET_i2s_full (&my_full_id));
  if (show_conns || monitor_connections)
  {
    mh = GNUNET_CORE_monitor_start (cfg, &monitor_cb, NULL);
    if (NULL == mh)
    {
      fprintf (stderr, "%s", _ ("Failed to connect to CORE service!\n"));
      ret = 1;
      GNUNET_SCHEDULER_shutdown ();
    }
    return;
  }
}


/**
 * The main function to obtain peer information from CORE.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  int res;
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_flag (
      'm',
      "monitor",
      gettext_noop (
        "provide information about all current connections (continuously)"),
      &monitor_connections),
    GNUNET_GETOPT_option_flag (
      'i',
      "show-identity",
      gettext_noop (
        "Show our current peer identity"
        ),
      &show_pid),
    GNUNET_GETOPT_option_flag (
      's',
      "connection-status",
      gettext_noop (
        "Show current connections"
        ),
      &show_conns),
    GNUNET_GETOPT_OPTION_END };

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;
  res = GNUNET_PROGRAM_run (argc,
                            argv,
                            "gnunet-core",
                            gettext_noop (
                              "Print information about connected peers."),
                            options,
                            &run,
                            NULL);

  GNUNET_free_nz ((void *) argv);
  if (GNUNET_OK == res)
    return ret;
  return 1;
}


/* end of gnunet-core.c */
